File: leetcode_cn/146/146.py
from typing import Dict, Optional


class Node:

    def __init__(self, _key=None, _value=None, _pre=None, _next=None):
        self.key = _key
        self.value = _value
        self.pre = _pre
        self.next = _next

    def __repr__(self):
        return f"<{self.key},{self.value}>{self.next}"


class LRUCache:

    def __init__(self, capacity: int):
        self._cache: Dict[int, Node] = dict()
        self.head: Node = None
        self.tail: Node = None
        self.capacity = capacity

    def __repr__(self):
        return f"<{self.head}>"

    def _update(self, key: int, value: Optional[int] = None):
        if value is not None:
            self._cache[key].value = value
        node = self._cache[key]

        # remove this node
        if node.pre is None:
            return
        else:
            pre = node.pre
        if node.next is not None:
            node.next.pre = pre
        else:
            self.tail = pre
        pre.next = node.next
        # print(f"after rm: {self}")

        # move this node to head
        node.pre = None
        node.next = self.head
        self.head.pre = node
        self.head = node
        # print(f"move to head: {self}")

    def _append(self, key: int, value: int):
        if self.tail is None:
            self.tail = Node(key, value)
            self.head = self.tail
        else:
            self.tail.next = Node(key, value, self.tail)
            self.tail = self.tail.next
        self._cache[key] = self.tail

    def _pop(self) -> Node:
        if self.tail is None:
            return
        key = self.tail.key
        node = self._cache[key]
        self._cache.pop(key)
        if node.pre is None:
            self.head = None
            self.tail = None
        else:
            self.tail = node.pre
            self.tail.next = None
        node.next = None
        node.pre = None
        return node

    def get(self, key: int) -> int:
        if key in self._cache:
            self._update(key)
            res = self._cache[key].value
        else:
            res = -1
        print(self)
        return res

    def put(self, key: int, value: int) -> None:
        if key in self._cache:
            self._update(key, value)
        else:
            if len(self._cache) == self.capacity:
                self._pop()
            self._append(key, value)
            self._update(key)
        print(self)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: leetcode_cn/146/146.cc
#include <iostream>
#include <unordered_map>
using namespace std;

class DLinkedNode {
public:
  int key, value;
  DLinkedNode *prev;
  DLinkedNode *next;
  DLinkedNode() : key(0), value(0), prev(nullptr), next(nullptr) {}
  DLinkedNode(int _key, int _value)
      : key(_key), value(_value), prev(nullptr), next(nullptr) {}
  DLinkedNode(int _key, int _value, DLinkedNode *_prev, DLinkedNode *_next)
      : key(_key), value(_value), prev(_prev), next(_next) {}
  void show() {
    cout << "<" << key << "," << value << ">,";
    if (next) {
      next->show();
    }
  }
};

class LRUCache {
private:
  unordered_map<int, DLinkedNode *> cache;
  int size;
  int cap;
  DLinkedNode *head;
  DLinkedNode *tail;

public:
  LRUCache(int capacity) {
    head = new DLinkedNode();
    tail = new DLinkedNode();
    head->next = tail;
    tail->prev = head;
    size = 0;
    cap = capacity;
  }

  ~LRUCache() {
    delete head;
    delete tail;
  }

  void show() {
    head->show();
    cout << endl;
  }

  void update(int key, int value) {
    DLinkedNode *node = cache[key];
    node->value = value;

    node->prev->next = node->next;
    node->next->prev = node->prev;

    node->next = head->next;
    node->prev = head;
    head->next->prev = node;
    head->next = node;
  }

  void update(int key) { update(key, cache[key]->value); }

  void add2Head(int key, int value) {
    DLinkedNode *node = new DLinkedNode(key, value, head, head->next);
    cache[key] = node;
    head->next->prev = node;
    head->next = node;
    size++;
  }

  void rmTail() {
    if (size == 0) {
      return;
    }
    size--;
    DLinkedNode *node = tail->prev;
    node->prev->next = tail;
    tail->prev = node->prev;
    cache.erase(node->key);
    delete node;
  }

  int get(int key) {
    if (!cache.count(key)) {
      return -1;
    }
    update(key);
    // show();
    return cache[key]->value;
  }

  void put(int key, int value) {
    if (cache.count(key)) {
      update(key, value);
    } else {
      add2Head(key, value);
      if (size > cap) {
        rmTail();
      }
    }
    // show();
  }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
